collector/streaming.go

```diff
@@ -0,0 +1,256 @@
+	for _, server := range nc.servers {
+		var resp StreamingServerz
+		if err := getMetricURL(nc.httpClient, server.URL, &resp); err != nil {
+			Debugf("ignoring server %s: %v", server.ID, err)
+			continue
+		}
+
+		ch <- prometheus.MustNewConstMetric(nc.bytesTotal, prometheus.CounterValue, float64(resp.TotalBytes), server.ID)
+		ch <- prometheus.MustNewConstMetric(nc.msgsTotal, prometheus.CounterValue, float64(resp.TotalMsgs), server.ID)
+		ch <- prometheus.MustNewConstMetric(nc.channels, prometheus.CounterValue, float64(resp.Channels), server.ID)
+		ch <- prometheus.MustNewConstMetric(nc.subs, prometheus.CounterValue, float64(resp.Subscriptions), server.ID)
+		ch <- prometheus.MustNewConstMetric(nc.clients, prometheus.CounterValue, float64(resp.Clients), server.ID)
+	}
+}
+
+type channelsCollector struct {
+	sync.Mutex
+
+	httpClient *http.Client
+	servers    []*CollectedServer
+
+	chanBytesTotal   *prometheus.Desc
+	chanMsgsTotal    *prometheus.Desc
+	chanLastSeq      *prometheus.Desc
+	subsLastSent     *prometheus.Desc
+	subsPendingCount *prometheus.Desc
+	subsMaxInFlight  *prometheus.Desc
+}
+
+func newChannelsCollector(servers []*CollectedServer) prometheus.Collector {
+	nc := &channelsCollector{
+		httpClient: http.DefaultClient,
+		chanBytesTotal: prometheus.NewDesc(
+			prometheus.BuildFQName("nss", "chan", "bytes_total"),
+			"Total of bytes",
+			[]string{"server", "channel"},
+			nil,
+		),
+		chanMsgsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName("nss", "chan", "msgs_total"),
+			"Total of messages",
+			[]string{"server", "channel"},
+			nil,
+		),
+		chanLastSeq: prometheus.NewDesc(
+			prometheus.BuildFQName("nss", "chan", "last_seq"),
+			"Last seq",
+			[]string{"server", "channel"},
+			nil,
+		),
+		subsLastSent: prometheus.NewDesc(
+			prometheus.BuildFQName("nss", "chan", "subs_last_sent"),
+			"Last message sent",
+			[]string{"server", "channel", "client_id"},
+			nil,
+		),
+		subsPendingCount: prometheus.NewDesc(
+			prometheus.BuildFQName("nss", "chan", "subs_pending_count"),
+			"Pending message count",
+			[]string{"server", "channel", "client_id"},
+			nil,
+		),
+		subsMaxInFlight: prometheus.NewDesc(
+			prometheus.BuildFQName("nss", "chan", "subs_max_inflight"),
+			"Max in flight message count",
+			[]string{"server", "channel", "client_id"},
+			nil,
+		),
+	}
+
+	// create our own deep copy, and tweak the urls to be polled
+	// for this type of endpoint
+	nc.servers = make([]*CollectedServer, len(servers))
+	for i, s := range servers {
+		nc.servers[i] = &CollectedServer{
+			ID:  s.ID,
+			URL: s.URL + "/streaming/channelsz?subs=1",
+		}
+	}
+
+	return nc
+}
+
+func (nc *channelsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- nc.chanBytesTotal
+	ch <- nc.chanMsgsTotal
+	ch <- nc.chanLastSeq
+	ch <- nc.subsLastSent
+	ch <- nc.subsPendingCount
+	ch <- nc.subsMaxInFlight
+}
+
+func (nc *channelsCollector) Collect(ch chan<- prometheus.Metric) {
```

Comment:
I think at some point we'll want "clientz", but this is great for now.

Reply:
Had also looked at clientsz, but seems that we could get all the same data from using channelsz, for example these two carry the same information only difference is that it was fetched from different endpoints (clientsz and channelsz respectively):

```sh
nss_client_last_sent{channel="foo",client_id="client-id-0",server="http://127.0.0.1:8222"} 46162
nss_chan_subs_last_sent{channel="foo",client_id="client-id-0",server="http://127.0.0.1:8222"} 46162
```